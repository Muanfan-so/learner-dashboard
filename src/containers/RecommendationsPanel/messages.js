import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  recommendationPanelHeading: {
    id: 'recommendation.panel.heading',
    defaultMessage: 'Recommendations for you',
    description: 'Personalized recommendations panel heading.',
  },
  exploreCoursesButton: {
    id: 'explore.courses.button',
    defaultMessage: 'Explore courses',
    description: 'Explore button on personalized recommendations panel.',
  },
});

export default messages;
