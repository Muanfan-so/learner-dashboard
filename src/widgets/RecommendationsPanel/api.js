import { StrictDict } from 'utils';
import { get, stringifyUrl } from 'data/services/lms/utils';
import urls from 'data/services/lms/urls';

export const getFetchUrl = () => (`${urls.getApiUrl()}/learner_recommendations/courses/`);
export const apiKeys = StrictDict({ user: 'user' });

const fetchRecommendedCourses = () => get(stringifyUrl(getFetchUrl()));

export default {
  fetchRecommendedCourses,
};
