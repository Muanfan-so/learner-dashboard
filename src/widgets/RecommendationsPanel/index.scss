@import "@edx/paragon/scss/core/core";

// Hard coded, should we be using paragon equivalent?
.explore-courses-btn {
  padding-top: 16px;
}

.panel-background {
  background: $light-200;
}

// Height value is hard coded. Would we we want poeple to be able to change the height?
.recommendations-loading {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 7.813rem;
  background: $light-200;
}
