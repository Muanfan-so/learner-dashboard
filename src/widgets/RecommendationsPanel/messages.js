import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  recommendationsHeading: {
    id: 'RecommendationsPanel.recommendationsHeading',
    defaultMessage: 'Recommendations for you',
    description: 'Heading on recommendations panel with personalized recommendations',
  },
  popularCoursesHeading: {
    id: 'RecommendationsPanel.popularCoursesHeading',
    defaultMessage: 'Popular on edX',
    description: 'Heading on recommendations panel with general recommendations',
  },
  exploreCoursesButton: {
    id: 'RecommendationsPanel.exploreCoursesButton',
    defaultMessage: 'Explore courses',
    description: 'Button to explore more courses on recommendations panel',
  },
});

export default messages;
