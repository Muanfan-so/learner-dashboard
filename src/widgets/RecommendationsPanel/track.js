import { StrictDict } from 'utils';
import { createLinkTracker, createEventTracker } from 'data/services/segment/utils';

export const eventNames = StrictDict({
  findCoursesClicked: 'edx.bi.dashboard.find_courses_button.clicked',
  recommendedCourseClicked: 'edx.bi.user.recommended.course.click',
});

export const findCoursesClicked = (href) => createLinkTracker(
  createEventTracker(eventNames.findCoursesClicked, {
    pageName: 'learner_home',
    linkType: 'button',
    linkCategory: 'search_button',
  }),
  href,
);

export const recommendedCourseClicked = (courseKey, isPersonalized, href) => createLinkTracker(
  createEventTracker(eventNames.recommendedCourseClicked, {
    course_key: courseKey,
    is_personalized_recommendation: isPersonalized,
  }),
  href,
);

export default {
  findCoursesClicked,
  recommendedCourseClicked,
};
