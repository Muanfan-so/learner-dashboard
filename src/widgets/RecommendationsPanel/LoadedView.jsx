import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';
import { Search } from '@edx/paragon/icons';

import { hooks } from 'data/redux';
import CourseCard from './components/CourseCard';
import messages from './messages';

import './index.scss';

export const LoadedView = ({
  courses,
  isPersonalizedRecommendation,
  courseSearchClickTracker,
}) => {
  const { courseSearchUrl } = hooks.usePlatformSettingsData();
  const { formatMessage } = useIntl();

  return (
    <div className="p-4 w-100 panel-background">
      <h3 className="pb-2">{isPersonalizedRecommendation
        ? formatMessage(messages.recommendationsHeading) : formatMessage(messages.popularCoursesHeading)}
      </h3>
      <div>
        {courses.map((course) => (
          <CourseCard
            key={course.courseKey}
            course={course}
            isPersonalizedRecommendation={isPersonalizedRecommendation}
          />
        ))}
      </div>
      <div className="text-center explore-courses-btn">
        <Button
          variant="tertiary"
          iconBefore={Search}
          as="a"
          href={courseSearchUrl}
          onClick={courseSearchClickTracker}
        >
          {formatMessage(messages.exploreCoursesButton)}
        </Button>
      </div>
    </div>
  );
};

LoadedView.propTypes = {
  courses: PropTypes.arrayOf(PropTypes.shape({
    courseKey: PropTypes.string,
    title: PropTypes.string,
    logoImageUrl: PropTypes.string,
    marketingUrl: PropTypes.string,
  })).isRequired,
  isPersonalizedRecommendation: PropTypes.bool.isRequired,
  courseSearchClickTracker: PropTypes.func.isRequired,
};

export default LoadedView;
