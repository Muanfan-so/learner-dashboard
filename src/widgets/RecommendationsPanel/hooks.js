import React from 'react';

import { StrictDict } from 'utils';
import { RequestStates } from 'data/constants/requests';
import { handleEvent } from 'data/services/segment/utils';

import * as module from './hooks';
import api from './api';

export const searchCourseEventName = 'learner_home.widget.search_course';

export const state = StrictDict({
  requestState: (val) => React.useState(val), // eslint-disable-line
  data: (val) => React.useState(val), // eslint-disable-line
});

export const useFetchCourses = (setRequestState, setData) => {
  React.useEffect(() => {
    let isMounted = true;
    api.fetchRecommendedCourses().then((response) => {
      if (isMounted) {
        setRequestState(RequestStates.completed);
        setData(response);
      }
    }).catch(() => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    });
    return () => { isMounted = false; };
    /* eslint-disable */
  }, []);
};

export const useRecommendationPanelData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchCourses(setRequestState, setData);
  const courses = data.data?.courses || [];
  const isPersonalizedRecommendation = data.data?.isPersonalizedRecommendation || false;
  const courseSearchClickTracker = () => handleEvent(searchCourseEventName, {
    pageName: 'learner_home',
    linkType: 'button',
    linkCategory: 'search_button',
  });
  return {
    courses,
    isPersonalizedRecommendation,
    isLoaded: requestState === RequestStates.completed && courses.length > 0,
    isFailed: requestState === RequestStates.failed
      || (requestState === RequestStates.completed && courses.length === 0),
    isLoading: requestState === RequestStates.pending,
    courseSearchClickTracker,
  };
};

export default {
  useRecommendationPanelData,
};
