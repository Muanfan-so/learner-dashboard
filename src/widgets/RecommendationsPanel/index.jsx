import React from 'react';

import LookingForChallengeWidget from 'widgets/LookingForChallengeWidget';
import LoadingView from './LoadingView';
import LoadedView from './LoadedView';
import hooks from './hooks';

export const RecommendationsPanel = () => {
  const {
    courses,
    isFailed,
    isLoaded,
    isLoading,
    courseSearchClickTracker,
  } = hooks.useRecommendationPanelData();

  if (isLoading) {
    return (<LoadingView />);
  }
  if (isLoaded) {
    return (<LoadedView courses={courses} courseSearchClickTracker={courseSearchClickTracker} />);
  }
  if (isFailed) {
    return (<LookingForChallengeWidget courseSearchClickTracker={courseSearchClickTracker} />);
  }
  // default fallback
  return (<LookingForChallengeWidget courseSearchClickTracker={courseSearchClickTracker} />);
};

export default RecommendationsPanel;
