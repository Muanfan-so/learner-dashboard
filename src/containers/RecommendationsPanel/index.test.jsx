import React from 'react';
import { shallow } from 'enzyme';

import RecommendationsPanel from '.';

const recommendedCourses = [
  {
    courseKey: 'edX+cs50',
    title: 'Computer Science',
    logoImageUrl: 'https://edx.org/small.jpg',
    marketingUrl: 'https://edx.org',
  },
  {
    courseKey: 'edX+cs50x',
    title: 'Course Test',
    logoImageUrl: 'https://edx.org/small.jpg',
    marketingUrl: 'https://edx.org',
  },
];

describe('RecommendationsPanel snapshot', () => {
  test('no recommended courses', () => {
    expect(shallow(<RecommendationsPanel courses={[]} />)).toMatchSnapshot();
  });

  test('has 2 recommended courses', () => {
    expect(shallow(<RecommendationsPanel courses={recommendedCourses} />)).toMatchSnapshot();
  });
});
