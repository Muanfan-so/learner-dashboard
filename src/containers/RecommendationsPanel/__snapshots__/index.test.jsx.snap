// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationsPanel snapshot has 2 recommended courses 1`] = `
<div
  className="widget-sidebar p-4"
>
  <h3
    className="pb-2"
  >
    Recommendations for you
  </h3>
  <div>
    <CourseCard
      course={
        Object {
          "course_key": "edX+cs50",
          "logoImageUrl": "https://edx.org/small.jpg",
          "marketingUrl": "https://edx.org",
          "title": "Computer Science",
        }
      }
    />
    <CourseCard
      course={
        Object {
          "course_key": "edX+cs50x",
          "logoImageUrl": "https://edx.org/small.jpg",
          "marketingUrl": "https://edx.org",
          "title": "Course Test",
        }
      }
    />
  </div>
  <div
    className="text-center pt-2"
  >
    <Button
      onClick={[Function]}
      variant=""
    >
      Explore courses
    </Button>
  </div>
</div>
`;

exports[`RecommendationsPanel snapshot no recommended courses 1`] = `
<div
  className="widget-sidebar p-4"
>
  <h3
    className="pb-2"
  >
    Recommendations for you
  </h3>
  <div />
  <div
    className="text-center pt-2"
  >
    <Button
      onClick={[Function]}
      variant=""
    >
      Explore courses
    </Button>
  </div>
</div>
`;
